### PR TITLE
allow git merge unrelated histories

### LINKED ISSUE #4239: yo jhipster:upgrade fails on 3.7.1 to 3.8.0 upgrade on first upgrade
##### **Overview of the issue**

Tried `yo jhipster:upgrade` on a project generated with 3.7.1 that is under git (and also has a Remote on BitBucket, everything pushed). This was the first try to upgrade so "Specific steps executed on first upgrade" was in effect. Build with 3.7.1 went well, then the following error message came:

```
Successfully regenerated app with jhipster 3.7.1
Committed with message "Generated with JHipster 3.7.1"
Checked out branch "master"
Error jhipster:upgrade

ERROR! Unable to record current code has been generated with version 3.7.1:
 fatal: refusing to merge unrelated histories

```

##### **Motivation for or Use Case**

Typical use case.

##### **Reproduce the error**

run yo jhipster:upgrade

##### **Related issues**

none with the "refusing to merge unrelated histories" message

##### **Suggest a Fix**

none

##### **JHipster Version(s)**

3.7.1

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**

{
  "generator-jhipster": {
    "jhipsterVersion": "3.7.1",
    "baseName": "LLMS",
    "packageName": "com.pannonrex.llms",
    "packageFolder": "com/pannonrex/llms",
    "serverPort": "8080",
    "authenticationType": "session",
    "hibernateCache": "ehcache",
    "clusteredHttpSession": false,
    "websocket": false,
    "databaseType": "sql",
    "devDatabaseType": "postgresql",
    "prodDatabaseType": "postgresql",
    "searchEngine": false,
    "buildTool": "maven",
    "enableSocialSignIn": false,
    "rememberMeKey": "...",
    "useSass": false,
    "applicationType": "monolith",
    "testFrameworks": [
      "gatling"
    ],
    "jhiPrefix": "jhi",
    "enableTranslation": true,
    "nativeLanguage": "en",
    "languages": [
      "en",
      "de",
      "hu"
    ]
  }
}

##### **Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

n/a

##### **Browsers and Operating System**

Chorme 53.0.2785.116 m (64-bit) (latest) / Win10

### PR DESCRIPTION
**Pull request by mehmet-aslan**
fix #4239 

Git does not allow merging two branches that have no common base by default since version 2.9.0 ([git release changelog](https://github.com/git/git/blob/master/Documentation/RelNotes/2.9.0.txt#L58-L68)). This prevents the merge commit that is made only on the first upgrade as the jhipster_upgrade branch is orphan.

I added git version check so if git version is 2.9.0 or newer merge could be made with --allow-unrelated-histories option.

### CODE REVIEW

**Comment by jdubois** [comment]
This is incredible!!!! I need to check, but THANK YOU as I would have a really hard time finding this.

File: generators/upgrade/index.js [hunk 1]
```diff
@@ -145,12 +146,26 @@ module.exports = UpgradeGenerator.extend({
 
         prepareUpgradeBranch: function() {
             var done = this.async();
+            var getGitVersion = function(callback) {
```
Comment (deepu105) [inline generators/upgrade/index.js]: I think its high time we start using arrow functions so lets start doing that.
we can avoid the doubel bind by doing
```
var getGitVersion = (callback) => {
    this.gitExec(['--version'], (code, msg)  => {
           callback(String(msg.match(/([0-9]+\.[0-9]+\.[0-9]+)/g)));
    });
};
```

**Comment by steelcoder** [comment]
Same issue for me (3.12.2 -> 4.10.2) with git 2.13.6 on Mac.

**Comment by ruddell** [comment]
If you have problems, please open a new issue and fill out the template completely.  Make sure to use the latest version of JHipster to upgrade.  It contains several `jhipster upgrade` bug fixes.

First upgrade your project's `generator-jhipster` dependency, then run `jhipster upgrade`
 - `yarn upgrade generator-jhipster` 
- `jhipster upgrade`

**Pull request merged**